### PR TITLE
Fixed #32317 -- Refactored loaddata command to make it extensible.

### PR DESCRIPTION
**Pull request by wkschwartz**
This refactors `loaddata` without changing the inputs or outputs of existing methods or introducing new public methods.

This will need to be squashed before merging, but I've left the commits separate for initial review. Some of the changes, such as 7385a56, look like they change semantics (but they don't!), and I wanted to include a commit message that justifies the change.

https://code.djangoproject.com/ticket/32317

### CODE REVIEW

File: django/core/management/commands/loaddata.py
Comment (shreyasravi) [inline django/core/management/commands/loaddata.py:103]: Does this need a short explanation now that it's factored out?

File: django/core/management/commands/loaddata.py
Comment (shreyasravi) [inline django/core/management/commands/loaddata.py:88]: Same comment as below

File: django/core/management/commands/loaddata.py
Comment (shreyasravi) [inline django/core/management/commands/loaddata.py:159]: May be an easier change for django reviewers if you keep this block in the same format as the one you replaced

File: django/core/management/commands/loaddata.py
Comment (shreyasravi) [inline django/core/management/commands/loaddata.py:162]: Same comment here as above comments. I'm guessing style wise though it's fine to leave out a docstring here as long as the name is informative enough (which I think all of them are).

**Review by shreyasravi — APPROVED** [review]
Looks all good.

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py:89]: Would it be better to discuss such a change in a separate ticket? I had intended this PR to be narrowly focused on code organization and readability. (Notice that this line is not actually new in this PR, but just [moved](https://github.com/django/django/pull/13842/files#diff-2e979ec7174d4cedb196eddcd10aca4dfa12f5041d89bd90d368b2953170d637L97)).

Comment (jacobtylerwalls) [reply django/core/management/commands/loaddata.py:89]: @smithdc1 perhaps you could leave a comment on #23919 with that TODO? My bias is to keep this pr smaller.

**Review by jacobtylerwalls — COMMENTED** [review]
<s>One question about translation, but</s> these changes look sensible. We'll want a more experienced contributor to chime in about how many of these changes would be welcome. If too much time wears on I can come back and give it a second look and mark it RFC to hot-potato it.

File: django/core/management/commands/loaddata.py
Comment (jacobtylerwalls) [reply django/core/management/commands/loaddata.py:159]: Then again, maybe nobody makes an effort to translate strings in the management commands, so maybe someone who knows can enlighten me.

Comment (jacobtylerwalls) [reply django/core/management/commands/loaddata.py:159]: I'm gonna admit to being criminally undercaffeinated when I started down this path. Disregard! ✔️

Comment (wkschwartz) [reply django/core/management/commands/loaddata.py:159]: > I'm gonna admit to being criminally undercaffeinated when I started down this path.

Relatable!

> Disregard! ✔️

I've never worked with translations before, but I wouldn't want to make things harder on people who use them. So just to convince myself that "disregard" is the right solution: I think `gettext` would need to be called somewhere for this string to be eligible for translation, and `gettext` is not even imported into the module.

File: django/core/management/commands/loaddata.py
Comment (pauloxnet) [inline django/core/management/commands/loaddata.py:23]: This format can be more clear:
```python
try:
    import bz2
except ImportError:
    has_bz2 = False
else:
    has_bz2 = True
```

File: django/core/management/commands/loaddata.py
Comment (pauloxnet) [inline django/core/management/commands/loaddata.py:29]: Same as above

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py:23]: I personally prefer what's currently there. If a committer/Django member prefers @pauloxnet's form, I'll change it.

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py:104]: 463566e467ebf21e9feffeb8c8d4c62a7d6e80c2

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py:88]: 463566e467ebf21e9feffeb8c8d4c62a7d6e80c2

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py:164]: 463566e467ebf21e9feffeb8c8d4c62a7d6e80c2

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py]: I removed the change in a5a69cab1a081b210d53703d6c8cd69e78362a7e given that it also confused @jacobtylerwalls https://github.com/django/django/pull/13842/files#r585972060

Even if the code is okay, it confused two separate reviewers. That's not a sign of clear code.

**Comment by wkschwartz** [comment]
I believe I've responded to all the outstanding reviews so far. I know this needs to be squashed, but I'll wait until a Django team member reviews it first.

File: django/core/management/commands/loaddata.py
Comment (felixxm) [inline django/core/management/commands/loaddata.py:163]: IMO we should revert this change `fixture_file` is an absolute path, so it can contain directories with dots in names, this can cause issues in `parse_name()`.

**Review by felixxm — COMMENTED** [review]
@wkschwartz Thanks :+1:  I've checked the first 6 commits, reorganized them and pushed small edits (see #14393). I'm going to rebase this PR after merging #14393.

File: django/core/management/commands/loaddata.py
Comment (felixxm) [inline django/core/management/commands/loaddata.py:275]: I don't see much value in adding `_check_fixture_files_in_dir()`. I would move this to the `_find_fixture_files_in_dir()` or leave it in the main loop.

File: django/core/management/commands/loaddata.py
Comment (felixxm) [inline django/core/management/commands/loaddata.py:266]: We don't need to add the `_` prefix to the new methods. They are considered a private API even without it.

File: django/core/management/commands/loaddata.py
Comment (felixxm) [inline django/core/management/commands/loaddata.py:258]: `product()` handles iterables so we don't need to call `list()`:
```suggestion
        cmp_fmts = self.compression_formats if cmp_fmt is None else [cmp_fmt]
```

File: django/core/management/commands/loaddata.py
Comment (felixxm) [inline django/core/management/commands/loaddata.py:264]: As far as I'm aware we don't need to iterate twice over the same list:
```suggestion
        return { 
            '%s.%s' % (
                fixture_name,
                '.'.join([ext for ext in combo if ext]),
            ) for combo in product(databases, ser_fmts, cmp_fmts)
        }
```

**Comment by felixxm** [comment]
Rebased.

@wkschwartz Do you have time to keep working on this? If not I can push edits.

**Comment by wkschwartz** [comment]
> @wkschwartz Do you have time to keep working on this?

I'll try to take a look today.

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py]: Done in e83e4694ef21e2b4bb7573cb98c2bf56f97c05d0

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py]: Agreed. I hadn't done this originally because I just moved the code without editing so that it'd show up as moved in `git diff --color-moved`.

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py]: For some reason GitHub wouldn't let me hit "Commit suggestion", but 455aea843c7f46e8c451d10fb5db6eb6073d1e2a should be identical to your suggestion. As with your other suggestion, I hadn't originally done it to minimize the diff, but happy to now that you've reviewed.

File: django/core/management/commands/loaddata.py
Comment (wkschwartz) [reply django/core/management/commands/loaddata.py]: Done in c0a62ed4181ae12715eaa341f7a772d829bb9a17

**Comment by wkschwartz** [comment]
I think I've responded to everything, and CI is passing, so I'll rebase and squash now.

**Review by felixxm — COMMENTED** [review]
@wkschwartz Thanks for updates :+1:

**Pull request merged**
